src/python/pants/pantsd/service/scheduler_service.py

```diff
@@ -22,41 +24,70 @@ class SchedulerService(PantsService):
 
   QUEUE_SIZE = 64
 
-  def __init__(self, fs_event_service, legacy_graph_helper):
+  def __init__(self, fs_event_service, legacy_graph_helper, build_root, invalidation_globs):
     """
     :param FSEventService fs_event_service: An unstarted FSEventService instance for setting up
                                             filesystem event handlers.
     :param LegacyGraphHelper legacy_graph_helper: The LegacyGraphHelper instance for graph
                                                   construction.
+    :param str build_root: The current build root.
+    :param list invalidation_globs: A list of `globs` that when encountered in filesystem event
+                                    subscriptions will tear down the daemon.
     """
     super(SchedulerService, self).__init__()
     self._fs_event_service = fs_event_service
     self._graph_helper = legacy_graph_helper
+    self._invalidation_globs = invalidation_globs
+    self._build_root = build_root
 
     self._scheduler = legacy_graph_helper.scheduler
     self._logger = logging.getLogger(__name__)
     self._event_queue = Queue.Queue(maxsize=self.QUEUE_SIZE)
     self._watchman_is_running = threading.Event()
+    self._invalidating_files = set()
 
   @property
   def change_calculator(self):
     """Surfaces the change calculator."""
     return self._graph_helper.change_calculator
 
+  @staticmethod
+  def _combined_invalidating_fileset_from_globs(glob_strs, root):
+    return set.union(*(Fileset.globs(glob_str, root=root)() for glob_str in glob_strs))
+
   def setup(self, lifecycle_lock, fork_lock):
     """Service setup."""
     super(SchedulerService, self).setup(lifecycle_lock, fork_lock)
     # Register filesystem event handlers on an FSEventService instance.
     self._fs_event_service.register_all_files_handler(self._enqueue_fs_event)
 
+    # N.B. We compute this combined set eagerly at launch with an assumption that files
```

Comment:
So, I think that this will work. It has some weaknesses related to symlinks I think, but that might be ok for the PYTHONPATH usecase.

Another approach that would be a bit more bulletproof, would be to request a `Snapshot` for the `PathGlobs` represented by the option, and to then recompute that `Snapshot` after _every_ file invalidation event (and it would need to be every event, because the `Snapshot` does not tell you which symlinks it traversed... it just guarantees that it does), to determine whether the `Snapshot`'s digest had changed. If the digest hadn't changed, this would take a few milliseconds... if it had changed, it would take a bit longer, but not much.

It may not be 100% general though, because if any of the globs represent symlinks outside of the working copy, we'd fail to cover them. Perhaps that requirement is fine?

Reply:
sounds good, but I think I'm going to punt on the `Snapshot` idea for this PR and tackle as a follow-up.